### ISSUE TITLE
Amesos2:  Confusing output when enabling MUMPS

### ISSUE THREAD

**Issue by modkin** [trilinos/Trilinos#7536]
## Bug Report
@trilinos/amesos2

### Description
When enabling the amesos2 package and the TPL MUMPS I get the following output during the cmake run:
`Setting Amesos2_ENABLE_MUMPS=ON since TPL_ENABLE_MUMPS=ON`
However, after cmake finished the variable `Amesos2_ENABLE_MUMPS` is `OFF`, and also Amesos2 was configured without MUMPS support.
This can be fixed by adding `-DAmesos2_ENABLE_MUMPS` manually to the cmake command but it is pretty confusing.
I came across this problem when installing trilinos with the package manager spack where this also occurs.

### Steps to Reproduce
1. SHA1:  55a75997332636a28afc9db1aee4ae46fe8d93e7 (release 12.18.1)
2. Cmake Call: `cmake .. -DTrilinos_ENABLE_Amesos2=ON -DTPL_ENABLE_MUMPS=ON`

**Comment by cgcgcg** [comment]
@trilinos/amesos2

**Comment by srajama1** [comment]
This might be a simple thing where we set -DAmesos2_ENABLE_MUMPS when TPL_ENABLE_MUMPS=ON. I thought TriBITS took care of this stuff for us. @bartlettroscoe

**Comment by bartlettroscoe** [comment]
> However, after cmake finished the variable Amesos2_ENABLE_MUMPS is OFF

@modkin, where did you see that?  What matters is what gets set in the generated config.h file.

Note that enable/disable does **not** change the value of CMake cache vars.  It only sets them in non-cache project-level vars.

@srajama1, is there no automated testing of Amesos2 with MUMPS?

**Comment by bartlettroscoe** [comment]
> Note that enable/disable does not change the value of CMake cache vars. It only sets them in non-cache project-level vars.

See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#tribits-auto-enables-disables-done-using-non-cache-local-variables

**Comment by modkin** [comment]
> @modkin, where did you see that? What matters is what gets set in the generated config.h file.

`Amesos2_ENABLE_MUMPS=OFF` is in the CMakeCache.txt. 
Also the cmake variable `HAVE_AMESOS2_MUMPS` variable is off. 
In `amesos2_config.h` there is only an commented out line:
`/* #undef HAVE_AMESOS2_MUMPS */`

I am aware that this does not change cached variables. I ran the cmake command in an empty build directory.

**Comment by srajama1** [comment]
@bartlettroscoe : MUMPS TPL is not supported by SEMS, so no. @trilinos/framework another TPL to test.

**Comment by bartlettroscoe** [comment]
@modkin, by any chance are you seeing the following in the CMake STDOUT:

```
Processing enabled package: Amesos2 (Libs, Tests, Examples)
CMake Warning at packages/amesos2/CMakeLists.txt:16 (MESSAGE):
  *****Amesos2 will not provide MUMPS Support with Tpetra_INST_INT_INT
  OFF.*****
```

?

I produced this on my local machine with:

```
$ rm -r CMake* *.out ;

$ cmake -DTrilinos_ENABLE_Amesos2=ON -DTPL_ENABLE_MUMPS=ON \
  -DTPL_MUMPS_INCLUDE_DIRS=/usr/include -DTPL_MUMPS_LIBRARIES=/usr/lib64/libblas.so \
  ~/Trilinos.base/Trilinos \
  &> configure.out
```

It seems that `Tpetra_INST_INT_INT` is `OFF` by default.   If you configure with:

But I am also seeing that the logic in:

* `Trilinos/packages/amesos2/CMakeLists.txt`

is incorrect.  It has:

```
TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_MUMPS
  HAVE_AMESOS2_MUMPS
  "Enable MUMPS in Amesos2"
  OFF
)
```

but it should have nothing at all about MUMPS or any other optional TPL support.  See:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

After I removed that block of code in the patch:

```
diff --git a/packages/amesos2/CMakeLists.txt b/packages/amesos2/CMakeLists.txt
index 2f4acdc..351b524 100644
--- a/packages/amesos2/CMakeLists.txt
+++ b/packages/amesos2/CMakeLists.txt
@@ -107,12 +107,6 @@ TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_ShyLU_NodeBasker
   ${Amesos2_HAVE_SHYLUBASKER}
 )
 
-TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_MUMPS
-  HAVE_AMESOS2_MUMPS
-  "Enable MUMPS in Amesos2"
-  OFF
-)
-
 ASSERT_DEFINED(TPL_ENABLE_METIS)
 TRIBITS_ADD_OPTION_AND_DEFINE(${PACKAGE_NAME}_ENABLE_METIS
   HAVE_AMESOS2_METIS
```

and configure with:

```
$ rm -r CMake* *.out ;

$ cmake -DTrilinos_ENABLE_Amesos2=ON -DTpetra_INST_INT_INT=ON -DTPL_ENABLE_MUMPS=ON \
  -DTPL_MUMPS_INCLUDE_DIRS=/usr/include -DTPL_MUMPS_LIBRARIES=/usr/lib64/libblas.so \
  ~/Trilinos.base/Trilinos \
  &> configure.out
```

it works fine as shown by:

```
$ grep MUMPS packages/amesos2/src/Amesos2_config.h 
/* Defined if you want to enable amesos2-MUMPS */
#define HAVE_AMESOS2_MUMPS
```

Looking over this more carefully, more of the logic in the file `Trilinos/packages/amesos2/CMakeLists.txt` is incorrect:

* It is defining vars like `Amesos2_ENABLE_CUSPARSE` and even through no dependency on that TPL is even declared in the file `Trilinos/packages/amesos2/cmake/Dependencies.cmake`.
* There are options defined like `Amesos2_ENABLE_LAPACK

What is strange is that many of the other TPLs being used by Amesos2 that are actually being tested in PR testing like METIS and ParMETIS are following the instructions outlined in:

* https://tribits.org/doc/TribitsDevelopersGuide.html#how-to-add-a-new-tribits-tpl-dependency

----

NOTE: Setting `TPL_MUMPS_INCLUDE_DIRS` and `TPL_MUMPS_LIBRARIES` to some dummy value bypasses checking and allows you to complete the configure stage even if you don't have MUMPS installed on your system (which I don't).

**Comment by bartlettroscoe** [comment]
@modkin, the fix in Amesos2 for this should be in PR #7562.  If you pull the branch in that PR and configure with  `-DTrilinos_ENABLE_Amesos2=ON -DTpetra_INST_INT_INT=ON -DTPL_ENABLE_MUMPS=ON` you should see:

```
$ grep MUMPS packages/amesos2/src/Amesos2_config.h 
/* Defined if you want to enable amesos2-MUMPS */
#define HAVE_AMESOS2_MUMPS
```

@srajama1, the @trilinos/amesos2 developers might consider tweaking how MUMPS support works in Amesos2.  The approach of allowing the output to say `Amesos2_ENABLE_MUMPS=ON` but not actually enabling support for MUMPS in Amesos2 is confusing.

What I would suggested is the following:

* Add logic to `Trilinos/cmake/RepositoryDepencenciesSetup.cmake` to set `set(Amesos2_ENABLE_MUMPS OFF)` if `Tpetra_INST_INT_INT` is not enabled.

* If the user does happen to explicitly set `=DAmesos2_ENABLE_MUMPS=ON`  but `Tpetra_INST_INT_INT` is not enabled, then error out in the `amesos2/CMakeLists.txt` file with a good
message to make it clear why.  (Warnings are just ignored by users.)

This is similar to logic you will find in  `Trilinos/cmake/RepositoryDepencenciesSetup.cmake` for `STKBalance`.

I can post a PR that makes these changes if you are interested.

**Comment by bartlettroscoe** [comment]
Sorry, I did not mean to close.

**Comment by modkin** [comment]
I can confirm that PR #7562 fixes the problem. With the same cmake command I originally used `HAVE_AMESOS2_MUMPS` is now defined.
The default for `Tpetra_INST_INT_INT` seams to have change only recently. On the `trilinos-release-12-18-branch` branch which I was using it is enabled by default.
Thank you very much for the help.
This can be closed when #7562 is merged.

**Comment by jhux2** [comment]
#7562 has been merged.
